core/src/main/java/kafka/log/remote/quota/RLMQuotaManager.java

```diff
@@ -0,0 +1,117 @@
+ * limitations under the License.
+ */
+package kafka.log.remote.quota;
+
+import kafka.server.QuotaType;
+import kafka.server.SensorAccess;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.MetricConfig;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.Quota;
+import org.apache.kafka.common.metrics.QuotaViolationException;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.SimpleRate;
+import org.apache.kafka.common.utils.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.runtime.BoxedUnit;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class RLMQuotaManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(RLMQuotaManager.class);
+
+    private final RLMQuotaManagerConfig config;
+    private final Metrics metrics;
+    private final QuotaType quotaType;
+    private final String description;
+    private final Time time;
+
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock();
+    private final SensorAccess sensorAccess;
+    private Quota quota;
+
+    public RLMQuotaManager(RLMQuotaManagerConfig config, Metrics metrics, QuotaType quotaType, String description, Time time) {
+        this.config = config;
+        this.metrics = metrics;
+        this.quotaType = quotaType;
+        this.description = description;
+        this.time = time;
+
+        this.quota = new Quota(config.getQuotaBytesPerSecond(), true);
+        this.sensorAccess = new SensorAccess(lock, metrics);
+    }
+
+    public void updateQuota(Quota newQuota) {
+        lock.writeLock().lock();
+        try {
+            this.quota = newQuota;
+
+            Map<MetricName, KafkaMetric> allMetrics = metrics.metrics();
+            MetricName quotaMetricName = metricName();
+            KafkaMetric metric = allMetrics.get(quotaMetricName);
+            if (metric != null) {
+                LOGGER.warn("Sensor for quota-id {} already exists. Setting quota to {} in MetricConfig", quotaMetricName, newQuota);
+                metric.config(getQuotaMetricConfig(newQuota));
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    public boolean isQuotaExceeded() {
+        Sensor sensorInstance = sensor();
+        try {
+            sensorInstance.checkQuotas();
+        } catch (QuotaViolationException qve) {
+            LOGGER.debug("Quota violated for sensor ({}), metric: ({}), metric-value: ({}), bound: ({})",
+                sensorInstance.name(), qve.metric().metricName(), qve.value(), qve.bound());
+            return true;
+        }
+        return false;
+    }
+
+    public void record(double value) {
+        sensor().record(value, time.milliseconds(), false);
```

Comment:
Why do we turn the quota checking off during `record()`? In other implementations like `ClientQuotaManager.recordAndGetThrottleTimeMs()`, we call `record()` by turning on quota checking and get back the amount of time to throttle.

Reply:
In KIP-956, we do not utilize the throttle time provided by the quota manager to regulate fetches and copies. For fetch operations, we initially verify quota availability before initiating the retrieval of remote data. If the quota is unavailable, our priority is to serve partitions requiring local data, rather than throttling the client. Therefore, we focus on fulfilling data requests for other partitions in the queue, eliminating the need for throttle time in fetch operations.

Similarly, when a RLM Task attempts to copy a segment, it first checks if the write quota is available. If the quota is not available, the thread waits until the quota becomes available. As a result, we do not require throttle time for copies either.